src/OpenFeature.Contrib.Providers.Flagd/InProcessResolver.cs

```diff
@@ -0,0 +1,276 @@
+using Grpc.Net.Client;
+using OpenFeature.Flagd.Grpc;
+using System.Net.Sockets;
+using System.Threading;
+using Grpc.Core;
+using Value = OpenFeature.Model.Value;
+
+namespace OpenFeature.Contrib.Providers.Flagd
+{
+    internal class InProcessResolver : Resolver
+    {
+        static int EventStreamRetryBaseBackoff = 1;
+
+        private readonly FlagSyncService.FlagSyncServiceClient _client;
+        private readonly JsonEvaluator _evaluator;
+
+        private readonly Mutex _mtx;
+        private int _eventStreamRetries;
+        private int _eventStreamRetryBackoff = EventStreamRetryBaseBackoff;
+        private readonly FlagdConfig _config;
+
+        internal InProcessResolver(FlagdConfig config)
+        {
+            _config = config;
+            _client = BuildClient(config, channel => new FlagSyncService.FlagSyncServiceClient(channel));
+            _mtx = new Mutex();
+            _evaluator = new JsonEvaluator(config.SourceSelector);
+        }
+
+        internal InProcessResolver(FlagSyncService.FlagSyncServiceClient client, FlagdConfig config)
+        {
+            _config = config;
+            _client = client;
+            _mtx = new Mutex();
+            _evaluator = new JsonEvaluator(config.SourceSelector);
+        }
+
+
+        public void Init()
+        {
+            var handleEvents = new Thread(HandleEvents);
+            handleEvents.Start();
+        }
+
+        public void Shutdown()
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
+        {
+            return Task.FromResult(_evaluator.ResolveBooleanValue(flagKey, defaultValue, context));
+        }
+
+        public Task<ResolutionDetails<string>> ResolveStringValue(string flagKey, string defaultValue, EvaluationContext context = null)
+        {
+            return Task.FromResult(_evaluator.ResolveStringValue(flagKey, defaultValue, context));
+        }
+
+        public Task<ResolutionDetails<int>> ResolveIntegerValue(string flagKey, int defaultValue, EvaluationContext context = null)
+        {
+            return Task.FromResult(_evaluator.ResolveIntegerValue(flagKey, defaultValue, context));
+        }
+
+        public Task<ResolutionDetails<double>> ResolveDoubleValue(string flagKey, double defaultValue, EvaluationContext context = null)
+        {
+            return Task.FromResult(_evaluator.ResolveDoubleValue(flagKey, defaultValue, context));
+        }
+
+        public Task<ResolutionDetails<Value>> ResolveStructureValue(string flagKey, Value defaultValue, EvaluationContext context = null)
+        {
+            return Task.FromResult(_evaluator.ResolveStructureValue(flagKey, defaultValue, context));
+        }
+
+        private async void HandleEvents()
+        {
+            while (_eventStreamRetries < _config.MaxEventStreamRetries)
```

Comment:
I think this needs to be improved. In-process provider is special as it only have a single flag sync mechanism (currently grpc stream). If we end reconnections after few retries, then flags store will go stale and this is not a desired state. Similar to what we do in other implementations, (for example Java [1]) we should use some exponential back-off (up to a limit) and continuously re-attempt to connect till we get a shutdown request. This should help with debugging as there will be continuous errors if something is wrong with connectivity (ex:- api address change) and reconnect and server fresh flags if problem goes away 

[1] - https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/storage/FlagStore.java#L88

Reply:
Also worth noting that some gRPC implementations have listener implementations so you can listen to the underlying connection state and restore the stream when the connection is restored. With this model, the stream retry options etc are used to control the parameters of the underlying connection retries, and the stream just reconnects accordingly.

Reply:
It looks like channel.WaitForStateChangedAsync could perhaps help us here, but its not available in all our targets :weary:

Reply:
@toddbaert  yeah, that's a good suggestion. And I hope unlike Java, .Net exposes configurations to control internal grpc reconnect configurations